### ISSUE TITLE
Code does not presently fit within bounds allowed by 32K v1.0 & 128K v2.0 bootloader

### ISSUE THREAD

**Issue by basilhussain** [basilhussain/stm8-bootloader-erase-write#3]
The code at present does not fit within the bounds set by the earliest 32K v1.0 and 128K v2.0 bootloaders. This is due to the lack of `NOP` padding on the erase/write routines,

For instance, the write routine entry point for 32K v1.0 is at `0x12F`, but the erase routine overlaps this boundary (extends to `0x13D`), and is truncated by the write routine. Similarly, the write routine entry point for 128K v2.0 is at `0x162`, but the erase routine extends to `0x175`, and so is also truncated by the write routine.

Perhaps something could be added to the build process which does a post-linking verification of the offsets and sizes, and issue a warning if any code overlaps. This could be done by reading the .map files generated by the linker.

**Comment by basilhussain** [comment]
A script has been added to the build process (check-seg-size.awk) that parses the .map files generated by the linker and checks whether any segments of interest overlap. If any overlaps are found, a warning is printed containing details. For example:

```
*** WARNING: Segment ERASE_SEG overlaps with following segment WRITE_SEG! ***
    ERASE_SEG ends @ 0x13c
    WRITE_SEG begins @ 0x12f
    Overlaps by 14 bytes
1 segment overlap(s) found
```

This will at least make it fairly obvious that the generated binaries for 32K v1.0 and 128K v2.0 are at present useless, until code size can be reduced.

**Comment by basilhussain** [comment]
Solved with commit 580c933. There are now no segment overlap warnings when building those versions.

For 32K v1.0, there is now 5 bytes margin between erase and write segments. For 128K v2.0 there is now 22 bytes margin.

The downside is that the common segment has expanded, and now extends to 0x253. We need to be aware that some of the smallest devices that feature a ROM bootloader only have 1KB of RAM but yet still a 512B stack. On these devices, the common segment will actually extend into the stack area. However, it shouldn't be a problem as the stack extends downwards and the erase/write routines use minimal amounts of stack space (a few bytes) during runtime (and it appears the ROM bootloader code uses none at all).
